### PR TITLE
189 ordinal scale domain item ordering

### PR DESCRIPTION
**Pull request by monfera**
CR: https://github.com/plotly/plotly.js/issues/189

Initial, minimal changeset for demonstrating the categorymode = 'array' option.

Test: 

```
Plotly.newPlot('embedded-graph', [
            {x: ['a','b',    'd','e','f'    ], y: [100,110,    130,140,150    ]},
            {x: ['a',    'c',    'e','f','g'], y: [101,    121,    141,151,161]}],
            {xaxis: {type: 'category', categorymode: 'array', categorylist: ['a','b','c','d','e','f','g']}}
        );
```

should output 
image

### CODE REVIEW

File: src/plots/cartesian/set_convert.js [hunk 1]
```diff
@@ -182,8 +182,9 @@ module.exports = function setConvert(ax) {
             // encounters them, ie all the categories from the
             // first data set, then all the ones from the second
             // that aren't in the first etc.
-            // TODO: sorting options - do the sorting
-            // progressively here as we insert?
+            // it is assumed that this function is being invoked in the
+            // already sorted category order; otherwise there would be
+            // a disconnect between the array and the index returned
```
Comment (etpinard) [inline src/plots/cartesian/set_convert.js:187]: @monfera can you elaborate on this? 

What else have you tried?

Comment (monfera) [inline src/plots/cartesian/set_convert.js:187]: @etpinard I tried things like this:
1. I've first tried the change _inside_ the suggested place (https://github.com/plotly/plotly.js/blob/bbb8205903cdd34ae019d5cd6a2c197e9a2550e7/src/plots/cartesian/set_convert.js#L179-L194): instead of pushing into the array, I inserted it to ensure orderedness (insertion sort). More accurately, for simplicity, I just sorted upon each insertion to try before doing too much work (paid attention to the fact that [].sort mutates the array). 
2. Then I also quickly tried doing the sort only _after_ the loop that invokes ax.d2c but the issue is, by that point we already have the indices, and a sort renders them obsolete.

Then I spent a bit of time better understanding concepts, the data flow and the various ways it can be invoked (e.g. plotting multiple lines; overplotting onto an already existing plot etc). Learnt that currently, the categorical X axis ordering is totally driven by the order in which the point tuples arrive and made sample plots. Which was the point at which I gave up on this point for modification and went for the lower hanging fruit of `array`. The other options will, I think, be more intrusive to the data flow.

Comment (monfera) [inline src/plots/cartesian/set_convert.js:187]: @etpinard maybe an example is helpful: The first vector has X cat. values `['a', 'c', 'd']`. This _is_ already sorted. Then a subsequent series is plotted in the same chart, with values for `['a', 'b', 'c', 'd']`. By this point, the function has returned indices such as `0, 1, 2` for `['a', 'c', 'd']`, respectively. Then in the next series comes a new point `'b'` correctly inserted at index `1` but the previously returned indices will become invalid and the plot will come out nonsensical (more precisely, having checked the plot, the way it came out was consistent with this mechanics).

Comment (monfera) [inline src/plots/cartesian/set_convert.js:187]: @etpinard I also considered doing a sorting a lot earlier (more upstream), but it would have issues: 1) the code should possibly reorder only _after_ it's established that it's a categorical axis (it can be user specified or determined heuristically); 2) by that point there are lots of places where `_fullData` etc. are persisted on objects in the trace order; 3) I believe it would be wrong to change the trace order just for this CR because of the snail trail example and the general possibility that the user does _not_ want to have plotly change the user input order (which is currently implicitly taken as the trace order).

So my current thought is that we need to do the axis tick sorting _downstream_, nearer the point which is responsible for the d3 data binding order for the axis ticks, and we have to accommodate for the possibility that new lines introduce new points that have to be inserted in lines that have already been added previously.

Comment (etpinard) [inline src/plots/cartesian/set_convert.js:187]: @monfera Thanks for this very detailed overview.

You've got me convinced, `ax._categories` can't be sorted from inside the `d2c` routine.

We need to fill in `ax._categories` earlier. I think within [`makeCalcdata`](https://github.com/plotly/plotly.js/blob/e03e9e0bd7ea3a6bed3b32bd784276e8e9eb70f1/src/plots/cartesian/set_convert.js#L208-L230) makes the most sense at the moment.

Comment (etpinard) [inline src/plots/cartesian/set_convert.js:187]: Oops. Actually, filling in `ax._categories` from `makeCalcdata` would lead to the same problems as you described above.

Comment (etpinard) [inline src/plots/cartesian/set_convert.js:187]: Filling in `ax._categories` within the [default step](https://github.com/plotly/plotly.js/blob/e03e9e0bd7ea3a6bed3b32bd784276e8e9eb70f1/src/plots/cartesian/axis_defaults.js) would definitely work (albeit with a slight performance hit).

We already need to loop all traces to check for box plot irregularities [here](https://github.com/plotly/plotly.js/blob/e03e9e0bd7ea3a6bed3b32bd784276e8e9eb70f1/src/plots/cartesian/axis_defaults.js#L156), so maybe you could fill in `ax._categories` within the same loop.

Comment (monfera) [inline src/plots/cartesian/set_convert.js:187]: @etpinard I'll look into it, thanks! Unrelated: I learnt that date axes take ISO strings `"2016-03-29"` as values. When I expressly indicate axis `type = 'category'` it behaves as categories and the axis ticks get rendered as plaintext `"2016-03-29"` rather than nicely formatted date. So I'm not yet grounded in the approach for coercion which was step #2 in your original comment.

Comment (monfera) [inline src/plots/cartesian/set_convert.js:187]: @etpinard I'll look into it more closely today, but on an initial look, the suggested place isn't as trivial (to me) and I'd put it slightly after this point, because 
- setAutoType bails if `ax.type!=='-'` and at the only place it's called from, there is an identical (redundant) check
- the loop in question depends on `isBoxWithoutPositionCoords` returning true
- the inside of the loop (now) is solely dedicated to building up the boxPositions
- if axis type `'category`' isn't explicitly specified, it won't be known until the subsequent call to `autoType()` at the end

Therefore I'm planning to put the logic _after_ having returned from setAutoType; the earliest point seems to be just before the call to `setConvert`. 

This way, we wouldn't reuse the suggested loop, but the loop over the traces probably doesn't have a measurable performance impact anyway (I'm assuming there usually aren't thousands or 10ks of trace lines).

I'll do some more tests to ensure I'm not overlooking something; just wanted to share current thinking.

Comment (etpinard) [inline src/plots/cartesian/set_convert.js:187]: Anywhere, in `plots/cartesian/axis_defaults.js` is fine at this stage. We'll fine tune the location for performance once the functionality is in a working state.

Comment (monfera) [inline src/plots/cartesian/set_convert.js:187]: @etpinard Thanks for the quick note, I'll go ahead.

File: test/jasmine/tests/calcdata_test.js [hunk 2]
```diff
@@ -28,4 +29,215 @@ describe('calculated data and points', function() {
...
+                expect(gd.calcdata[0][3].y).toEqual(13);
+                expect(gd.calcdata[0][4].y).toEqual(14);
+            });
+
+            it('should output categories in the given order if `trace` order is explicitly specified', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'trace'
+                    // Wouldn't it be preferred to supply a function and plotly would have several functions like this?
```
Comment (etpinard) [inline test/jasmine/tests/calcdata_test.js]: it would be nice :smirk: , but only for JS users. 

All plotly.js attributes must be JSON serializable so that folks using our [python](https://github.com/plotly/plotly.py) and [R](https://github.com/ropensci/plotly) libraries can have access to the same features.

File: test/jasmine/tests/calcdata_test.js [hunk 3]
```diff
@@ -28,4 +29,215 @@ describe('calculated data and points', function() {
...
+                    // E.g. it's easier for symbol completion (whereas there's no symbol completion on string config)
+                    // See arguments from Mike Bostock, highlighted in medium green here:
+                    // https://medium.com/@mbostock/what-makes-software-good-943557f8a488#eef9
+                    // Plus if it's a function, then users can roll their own.
+                    //
+                    // Also, if axis tick order is made configurable, shouldn't we make trace order configurable?
+                    // Trace order as in, if a line or curve is drawn through points, what's the trace sequence.
+                    // These are two orthogonal concepts. In this round, I'm assuming that the trace order is implied
+                    // by the order the {x,y} arrays are specified.
```
Comment (etpinard) [inline test/jasmine/tests/calcdata_test.js:58]: > In this round, I'm assuming that the trace order is implied by the order the {x,y} arrays are specified

... and you would be correct.

File: src/plot_api/plot_api.js [hunk 4]
```diff
@@ -851,10 +851,10 @@ function doCalcdata(gd) {
     fullLayout._piecolormap = {};
     fullLayout._piedefaultcolorcount = 0;
 
-    // delete category list, if there is one, so we start over
+    // initialize the category list, if there is one, so we start over
     // to be filled in later by ax.d2c
     for(i = 0; i < axList.length; i++) {
-        axList[i]._categories = [];
+        axList[i]._categories = axList[i]._initialCategories.slice();
```
Comment (etpinard) [inline src/plot_api/plot_api.js:857]: nice find.

File: src/plots/cartesian/layout_attributes.js [hunk 5]
```diff
@@ -446,6 +446,36 @@ module.exports = {
...
         ].join(' ')
     },
+    categorymode: {
+        valType: 'enumerated',
+        values: [
+            'trace', 'category ascending', 'category descending',
+            'value ascending', 'value descending','array'
+        ],
+        dflt: 'trace',
+        role: 'style',
```
Comment (etpinard) [inline src/plots/cartesian/layout_attributes.js]: `role: 'info'.`

By the way, the attribute `role` fields have no use in plotly.js. They are [stripped](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/tasks/util/compress_attributes.js#L31) out on the bundle step. They are only used in [plotly.py](https://plot.ly/python/matplotlib-to-plotly-tutorial/) where they indicate which attribute to strip when the `strip_style` flag is turned on.

File: src/plots/cartesian/layout_attributes.js [hunk 6]
```diff
@@ -446,6 +446,36 @@ module.exports = {
...
+            'Set `categorymode` to *value ascending* or *value descending* if order should be determined by the',
+            'numerical order of the values.',
+            'Set `categorymode` to *array* to derive the ordering from the attribute `categorylist`. If a category',
+            'is not found in the `categorylist` array, the sorting behavior for that attribute will be identical to',
+            'the *trace* mode. The unspecified categories will follow the categories in `categorylist`.'
+        ].join(' ')
+    },
+    categorylist: {
+        valType: 'data_array',
+        role: 'style',
```
Comment (etpinard) [inline src/plots/cartesian/layout_attributes.js]: same here `role: 'info'`

File: src/plots/cartesian/ordered_categories.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    return categorymode === 'array' ?
+
+        // just return a copy of the specified array ...
+        categorylist.slice() :
+
+        // ... or take the union of all encountered tick keys and sort them as specified
+        // (could be simplified with lodash-fp or ramda)
+        [].concat.apply([], data.map(function(d) {return d[axisLetter];}))
+            .filter(function(element, index, array) {return index === array.indexOf(element);})
+            .sort(({
```
Comment (etpinard) [inline src/plots/cartesian/ordered_categories.js]: We like using for loops in these situations. Although they may be not as nice to look at than these functional patterns, they are much [faster](https://jsperf.com/array-loop-vs-filter). :racehorse:

File: test/jasmine/tests/calcdata_test.js [hunk 8]
```diff
@@ -28,4 +29,215 @@ describe('calculated data and points', function() {
...
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'array',
+                    categorylist: ['b','a','x','c']
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(13);
+                expect(gd.calcdata[0][1].y).toEqual(11);
+                expect(gd.calcdata[0][2].y).toEqual(15);
+
```
Comment (etpinard) [inline test/jasmine/tests/calcdata_test.js:326]: Hmm. I think `gd.calcdata[0][2].y` should be `null` in this case and  `gd.calcdata[0][3].y` be equal to 15.

File: test/jasmine/tests/calcdata_test.js [hunk 9]
```diff
@@ -28,4 +29,215 @@ describe('calculated data and points', function() {
...
+            it('should output categories in explicitly supplied order even if not all categories are present', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'array',
+                    categorylist: ['y','b','x','a','d','z','e','c']
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(13);
```
Comment (etpinard) [inline test/jasmine/tests/calcdata_test.js]: Same here.

Specified categories in `categorylist` that don't have associated values in the `data` should be understood as `null`s.

Comment (monfera) [inline test/jasmine/tests/calcdata_test.js]: @etpinard Yes, totally agree with both these comments. The test cases came before the implementation, i.e. before some of my learning and haven't been updated.

File: test/jasmine/tests/calcdata_test.js [hunk 10]
```diff
@@ -28,4 +29,215 @@ describe('calculated data and points', function() {
...
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,null,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'value descending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(15);
+                expect(gd.calcdata[0][1].y).toEqual(14);
+                expect(gd.calcdata[0][2].y).toEqual(12);
+                expect(gd.calcdata[0][3].y).toEqual(11);
```
Comment (etpinard) [inline test/jasmine/tests/calcdata_test.js:183]: this is correct, the `null` items should disappear from the calcdata data. Similar to the current [behavior](http://codepen.io/etpinard/pen/NxZLvZ).

File: src/plots/cartesian/ordered_categories.js (hunk 7 above)
Comment (monfera) [inline src/plots/cartesian/ordered_categories.js]: @etpinard Thanks, it'll be a quick update from me.

File: test/jasmine/tests/calcdata_test.js [hunk 11]
```diff
@@ -28,4 +29,215 @@ describe('calculated data and points', function() {
             expect(gd.calcdata[0][3]).toEqual({ x: false, y: false});
         });
     });
+
+    xdescribe('category ordering', function() {
```
Comment (etpinard) [inline test/jasmine/tests/calcdata_test.js]: @monfera these tests are great. The test :tiger2: thanks you! 

Could you add a couple more cases where multiple traces with [shared](http://codepen.io/etpinard/pen/JXOMVZ) and/or [mutually exclusive](http://codepen.io/etpinard/pen/ZQxxWb) categories?

Comment (monfera) [inline test/jasmine/tests/calcdata_test.js]: @etpinard Thanks :-) Yes, I guess we'll have at least double of the number of test cases (corner cases etc.). Just didn't want to race ahead of your initial feedback and instead, I looked into the unrelated bug you sent my way. Thanks for these specific suggestions.

File: src/plots/cartesian/axis_defaults.js [hunk 12]
```diff
@@ -64,6 +65,10 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, coerce,
         }
     }
 
+    containerOut._initialCategories = axType === 'category' ?
+        orderedCategories(letter, containerIn.categorymode, containerIn.categorylist, options.data) :
```
Comment (etpinard) [inline src/plots/cartesian/axis_defaults.js:70]: We'll need to coerce `categorymode` and `cateogrylist` before sending them to `orderedCategories`.

Coercion is done in this scope by the [shortcut coerce function](https://github.com/monfera/plotly.js/blob/9c366ce959b604ca55c09ddf49b357e712fcf0eb/src/plots/cartesian/layout_defaults.js#L127) defined in `plots/cartesian/layout_defaults.js`. The shortcut coerce function is wrapper around [`Lib.coerce`](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/src/lib/coerce.js#L290-L312) which applies coercion rules based on the type of the user input. For example, an [enumerated attribute](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/src/lib/coerce.js#L34-L46) can only be set to the possible values listed in the [attributes definitions](https://github.com/monfera/plotly.js/pull/1/files#diff-0e41c2e162564438ff091d0ed6b5b455R451), if not, it gets coerced to the default value.

Moreover, we'll need to add logic around `categorymode` and `categorylist`. For instance, at the moment, `categorylist` only has an effect if `categorymode` is set to `array`. Meaning that if a user specifies `categorylist` without `categorymode`, the defaults categories will be shown even though it is clear that the user wanted custom categories. So, we can be smarter by overriding the `catogorymode` default if `categoylist` is an array, similar to what's being done currently for `tickmode` and `tickvals` in [`plots/cartesian/tick_value_defaults.js`](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/src/plots/cartesian/tick_value_defaults.js#L23-L27) (see tests [here](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/test/jasmine/tests/axes_test.js#L397-L490) for more info).

We may want to look up [`lib/nested_properly.js`](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/src/lib/nested_property.js) while familiarizing yourself with `Lib.coerce` (see its test cases [here](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/test/jasmine/tests/lib_test.js#L249-L436)).

Don't hesitate to ask me any questions about our coercion framework. Surprising, it may the trickiest part to understand about plotly.js.

File: test/jasmine/tests/calcdata_test.js [hunk 13]
```diff
@@ -75,11 +75,11 @@ describe('calculated data and points', function() {
...
-                expect(gd.calcdata[0][1].y).toEqual(13);
-                expect(gd.calcdata[0][2].y).toEqual(15);
-                expect(gd.calcdata[0][3].y).toEqual(14);
-                expect(gd.calcdata[0][4].y).toEqual(12);
+                expect(gd.calcdata[0][0]).toEqual(jasmine.objectContaining({x: 2, y: 15}));
+                expect(gd.calcdata[0][1]).toEqual(jasmine.objectContaining({x: 0, y: 11}));
+                expect(gd.calcdata[0][2]).toEqual(jasmine.objectContaining({x: 4, y: 12}));
+                expect(gd.calcdata[0][3]).toEqual(jasmine.objectContaining({x: 1, y: 13}));
+                expect(gd.calcdata[0][4]).toEqual(jasmine.objectContaining({x: 3, y: 14}));
             });
```
Comment (monfera) [inline test/jasmine/tests/calcdata_test.js:83]: @etpinard Just an FYI, since I started this CR with the test cases, my conception on where order would be present became outdated. As now I understand that gd.calcdata will keep the trace order, I'm updating test cases such that the original trace order is expected, and explicit checks on the x/y tuples are in place. It looks good to me but please tell me if you have an alternative suggestion.

Comment (etpinard) [inline test/jasmine/tests/calcdata_test.js:83]: nicely done.

File: test/jasmine/tests/calcdata_test.js (hunk 10 above)
Comment (monfera) [inline test/jasmine/tests/calcdata_test.js:183]: @etpinard Clarifications:
- When either the category or the value of a trace point is null, it gets mapped to {x: false, y: false} by the current production code, rather than a full disappearance, I haven't changed it.
- In case a category is enlisted in `categorylist` but has no point for it in the trace: the x values in the {x,y} index tuples will correctly 'jump over' the category which doesn't have a data point. However there's no artificially generated point for such a category so here as well, we don't exactly have a null value (simply, there won't be a tuple with x as a missing category).

As I'm looking at the generated plots as well, I found something unexpected: when the first N or last N categories in `categorylist` have no corresponding point(s) in the trace, they're chopped from the axis. My guess is that it is to do with the automatic zooming onto the actual range of the data, usually done in the case of continuous axes. 

I believe it's not the desired behavior for categorical axes. If it's indeed a range focus thing, then it's a preexisting condition that arose now because in the past, the axis ticks were driven by the trace points, so trivially, there used to be at least one point for each category. Should I work on solving it, or should we optimize for as quick merging of the baseline functionality as possible?

Here's the data:

```
Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
    type: 'category',
    categorymode: 'array',
    categorylist: ['s','y','b','x','a','d','z','e','c', 'q', 'k']
}});
```

and here's the chart: 
image

Comment (etpinard) [inline test/jasmine/tests/calcdata_test.js:183]: @monfera thanks for this writeup. You bring up some interesting points.

plotly.js has always been very data-focused. For example, data linked to `'x'` and/or `'y'` are used to _compute_ axis ranges by default. In other words, data can impact other graph attributes, but we like to keep data arrays immutable.

In the case where `categorylist` has more entries to `'x`', making the auto-range routine consider the orphan categories in `categorylist` would be equivalent to filling in `'x'` with those categories and the corresponding `'y'` items will `null`s. Note that null `'y'` items are taken into account in the auto-range routine (see [example](http://codepen.io/etpinard/pen/jqYBER)).

Therefore, I'd vote for the status quo, where orphan `categorylist` categories don't affect the auto-range routine. In the case where a user wants to see those orphan categories, adding (category, `null`) x/y pairs or setting `xaxis.range` explicitly would suffice.

Comment (monfera) [inline test/jasmine/tests/calcdata_test.js:183]: @etpinard Thanks for the answer, it means we don't need to worry about chopping off unused categories on the axis tails for now. I'm just adding a test case that renders the tick for a `null` value, it's working out fine.

File: test/jasmine/tests/calcdata_test.js [hunk 14]
```diff
@@ -30,7 +30,7 @@ describe('calculated data and points', function() {
         });
     });
 
-    xdescribe('category ordering', function() {
+    describe('category ordering', function() {
```
Comment (etpinard) [inline test/jasmine/tests/calcdata_test.js:33]: quick note: you can run [`npm run test-syntax`](https://github.com/plotly/plotly.js/blob/master/test/syntax_test.js) to check that you haven't forgotten any `fdescribe` or `xit` in your jasmine test suites.

File: test/jasmine/tests/calcdata_test.js [hunk 15]
```diff
@@ -236,6 +232,14 @@ describe('calculated data and points', function() {
...
                 expect(gd.calcdata[0][3]).toEqual(jasmine.objectContaining({x: 1, y: 13}));
                 expect(gd.calcdata[0][4]).toEqual(jasmine.objectContaining({x: 4, y: 14}));
+
+                // The auto-range feature currently eliminates unused category ticks on the left/right axis tails.
+                // The below test case reifies this current behavior, and checks proper order of categories kept.
+
+                var domTickTexts = Array.prototype.slice.call(document.querySelectorAll('g.xtick'))
+                    .map(function(e) {return e.__data__.text;});
+
+                expect(domTickTexts).toEqual(['b', 'x', 'a', 'd', 'z', 'e', 'c']);  // y, q and k has no data points
```
Comment (monfera) [inline test/jasmine/tests/calcdata_test.js:271]: @etpinard a few questions regarding this type of test:
1. I've used library-free querying of the DOM; I think it's good but let me know if I should import d3 for this purpose, or you could point out another test file as a pattern to follow.
2. I haven't yet bumped into axes/ticks tests of this nature, where we query the DOM, which is downstream of the 'viewModel' testing this file mostly has but upstream of image based testing. Do we need such DOM based testing, and are there preexisting DOM tests I should instead use that ensure e.g. proper category or tracing order?
3. If we need such tests, should I add similar ones for the other test cases in this file?

File: test/jasmine/tests/calcdata_test.js [hunk 16]
```diff
@@ -255,6 +259,11 @@ describe('calculated data and points', function() {
                 expect(gd.calcdata[0][2]).toEqual(jasmine.objectContaining({x: 6, y: 12}));
                 expect(gd.calcdata[0][3]).toEqual(jasmine.objectContaining({x: 1, y: 13}));
                 expect(gd.calcdata[0][4]).toEqual(jasmine.objectContaining({x: 4, y: 14}));
+
+                var domTickTexts = Array.prototype.slice.call(document.querySelectorAll('g.xtick'))
+                    .map(function(e) {return e.__data__.text;});
+
+                expect(domTickTexts).toEqual(['y', 'b', 'x', 'a', 'd', 'z', 'e', 'c']);  // q, k has no data; y is null
```
Comment (monfera) [inline test/jasmine/tests/calcdata_test.js:295]: @etpinard Same question here.

File: test/jasmine/tests/calcdata_test.js (hunk 14 above)
Comment (monfera) [inline test/jasmine/tests/calcdata_test.js:33]: @etpinard is there a recommended script that check pretty much everything? Should I ensure that each commit is free of lint, fdescribe etc. issues? I prefer granular commits but some processes (jasmine and image tests) take significant time and even lint/syntax is something to keep in mind. If commits must be 'clean' in master e.g. to aid `git bisect`, I can squash commits e.g. after a PR review or we could try the new github function which squashes on merge (https://github.com/blog/2141-squash-your-commits)?

Depending on the answer, may I use an existing script e.g. `npm test` that execute multiple aspects, or even make a new one with lint?

File: src/plots/cartesian/axis_defaults.js (hunk 12 above)
Comment (monfera) [inline src/plots/cartesian/axis_defaults.js:70]: @etpinard Looking into this now as it appears to be the last piece in this CR. My test cases will need an update once coercion is in place, to switch to the more data-driven default or override behavior.

Comment (monfera) [inline src/plots/cartesian/axis_defaults.js:70]: @etpinard tl;dr blocking question for you at the end.

As an aid to thinking about goals, here's a summary of what coercions I'm planning:
1. If `categorymode` is not one of those in the attributes definition, it'll be coerced to the default `trace` value.
2. No matter what `categorymode` is specified or coerced in item#1, if a `categorylist` is provided and it is an array, `categorymode` will be coerced into `array`.

... well it looks like the two things you listed :-) My original take was that `categorymode`, as a switch, drives the show in an unambiguous way, and the user can flip-flop among modes without worrying about removing and readding (or commenting/uncommenting) the `categorylist`  property. Even now I'm vague about the case where `categorymode` is `category ascending` and `categorylist` is also given. Moreover, it renders the `categorymode === 'array'` redundant (documentational at best) since the presence of `categorylist` takes precedence. Maybe we should even do away with `categorymode === 'array'`, making `categorymode` and `categorylist` mutually exclusive?

Taking it one step further, shouldn't we just have _one_ property, e.g. called `categoryorder` that can be `category ascending`, `category descending` or an array?

I'm now plunging into the basic coercion so we don't have an unexpected enumerative value but your input to the above question would be useful. It's not a big deal to unify the two properties and at least technically, it would be less ambiguous.

Comment (etpinard) [inline src/plots/cartesian/axis_defaults.js:70]: > My original take was that categorymode, as a switch, drives the show in an unambiguous way, and the user can flip-flop among modes without worrying about removing and readding (or commenting/uncommenting) the categorylist property.

... and you would be correct.

What _smart coercion_ is trying to remove redundancy. For example,

``` js
var layout: {
  xaxis: {
    categorylist: ['apples', 'bananas', 'clementines']
  }
};
```

should be considered the same as:

``` js
var layout: {
  xaxis: {
    categorymode: 'array',  // redundant !!!
    categorylist: ['apples', 'bananas', 'clementines']
  }
};
```

But, 

``` js
var layout: {
  xaxis: {
    categorymode: 'category descending',  // still used as a switch 
    categorylist: ['apples', 'bananas', 'clementines']  // not used in graph
  }
};
```

Comment (monfera) [inline src/plots/cartesian/axis_defaults.js:70]: @etpinard Thanks! <eom>

Comment (etpinard) [inline src/plots/cartesian/axis_defaults.js:70]: and @monfera I should add, patterns like :arrow_double_up: are common in plotly.js. 

Don't try to reinvent the wheel. As mentioned before, this [block](https://github.com/plotly/plotly.js/blob/fb874424d5b89c021ef6e5ebe0326433389d02ee/src/plots/cartesian/tick_value_defaults.js#L23-L27) should serve as a good guide for how `categorymode` and `categorylist` should interact.

File: test/jasmine/tests/axes_test.js [hunk 17]
```diff
@@ -7,8 +7,8 @@ var Color = require('@src/components/color');
 var handleTickValueDefaults = require('@src/plots/cartesian/tick_value_defaults');
 var Axes = PlotlyInternal.Axes;
 
-var createGraph = require('../assets/create_graph_div');
-var destroyGraph = require('../assets/destroy_graph_div');
+var createGraphDiv = require('../assets/create_graph_div');
+var destroyGraphDiv = require('../assets/destroy_graph_div');
```
Comment (monfera) [inline test/jasmine/tests/axes_test.js:12]: @etpinard I appended 'Div' to these, to be in sync with how it's usually called.

File: test/jasmine/tests/axes_test.js [hunk 18]
```diff
@@ -321,15 +321,137 @@ describe('Test axes', function() {
...
+                PlotlyInternal.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], {
+                    xaxis: {type: 'category', categorymode: 'invalid value', categorylist: ['b','a','d','e','c']}
+                });
+                expect(gd._fullLayout.xaxis.categorymode).toBe('array');
+            });
+
+        });
+
+    });
+
```
Comment (monfera) [inline test/jasmine/tests/axes_test.js:445]: @etpinard May I ask you to take a look at these test cases if they seem to be in line with what you wrote about categorymode coercion earlier.

File: test/jasmine/tests/axes_test.js (hunk 17 above)
Comment (etpinard) [inline test/jasmine/tests/axes_test.js:12]: nicely done.

File: src/plots/cartesian/category_mode_defaults.js [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+            coerce('categorymode');
+
+        } else {
+
+            // otherwise use the supplied mode, or the default one if unsupplied or invalid
+            coerce('categorymode');
+
+        }
+    }
+};
```
Comment (monfera) [inline src/plots/cartesian/category_mode_defaults.js:39]: @etpinard Is this a reasonable cut at the coercion?

File: test/jasmine/tests/axes_test.js (hunk 18 above)
Comment (etpinard) [inline test/jasmine/tests/axes_test.js:445]: Great. All the test cases you added (along with d16fe6b13a357b2e603caf6d7e376e5d1dea28b0) look great!

File: src/plots/cartesian/category_mode_defaults.js [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+module.exports = function handleCategoryModeDefaults(containerIn, containerOut, coerce) {
+
+    if(containerIn.type === 'category') {
```
Comment (etpinard) [inline src/plots/cartesian/category_mode_defaults.js]: we like early `return`s in case like these.

So,

``` js
if(containerIn.type !== 'category') return;
```

and flatten :arrow_double_down:

File: src/plots/cartesian/category_mode_defaults.js [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+
+        var validCategories = ['trace', 'category ascending', 'category descending', 'array'];
```
Comment (etpinard) [inline src/plots/cartesian/category_mode_defaults.js]: we should `require('./layout_attrbitues')` into this file and have:

``` js
var validCategories = layoutAttributes.categorymode.values;
```

to keep things DRY :palm_tree:

File: src/plots/cartesian/category_mode_defaults.js [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+
+        var properCategoryList = Array.isArray(containerIn.categorylist) && containerIn.categorylist.length > 0;
+
+        if(validCategories.indexOf(containerIn.categorymode) === -1 && properCategoryList) {
```
Comment (etpinard) [inline src/plots/cartesian/category_mode_defaults.js]: :+1: 

Coercing while keeping track of whether the user input in `containerIn` is valid was the idea behind [`Lib.coerce2`](https://github.com/plotly/plotly.js/blob/db9fd34251f099907837f4fc8c56e74faa13b9e1/src/lib/coerce.js#L321-L326). 

`Lib.coerce2` is not very well implemented at the moment, and I want to change its API to make more convenient.

So, no action needed here. This logic works great and is well tested.

File: src/plots/cartesian/category_mode_defaults.js (hunk 19 above)
Comment (etpinard) [inline src/plots/cartesian/category_mode_defaults.js:39]: nice job.

File: src/plots/cartesian/ordered_categories.js [hunk 23]
```diff
@@ -11,32 +11,44 @@
...
         }
     }
-    return categoryArray;
-}
 
-// flattenUniqueSort :: String -> Function -> [[String]] -> [String]
-function flattenUniqueSort(axisLetter, sortFunction, data) {
-    return flattenUnique(axisLetter, data).sort(sortFunction);
+    return categoryArray;
 }
```
Comment (monfera) [inline src/plots/cartesian/ordered_categories.js:53]: @etpinard I quite rewrote the ordering logic this evening on initial suspicion that image based cases may fail in presence of a large number of points, because of possible slowness due to the O(N) complexity. I switched to bisection. While it didn't make a difference, better scalability is good anyway. As it changed a good bit, it's useful if you know about it.

Comment (etpinard) [inline src/plots/cartesian/ordered_categories.js:53]: looks good :+1:
